### PR TITLE
added a ListenerAddr struct which "owns" a TCP port

### PR DESCRIPTION
**Pull request by pompon0**
The purpose of this is to avoid race conditions when assigning random TCP ports in tests. In test ListenerAddr is keeping a TCP socket bound to a port (but not listening). We utilize SO_REUSEPORT socket option to open a listener socket (of independent lifespan) on the same port.

### CODE REVIEW

File: chain/network/src/tcp.rs [hunk 1]
```diff
@@ -118,15 +118,74 @@ impl Stream {
...
+}
+
+impl ListenerAddr {
+    pub fn new(addr: std::net::SocketAddr) -> Self {
+        assert!(addr.port()!=0, "using an arbitrary port for the tcp::Listener is allowed only in tests and only via new_for_test() method");
+        Self { addr, guard: Arc::new(None) }
     }
 
+    /// TEST-ONLY: constructs a ListenerAddr owning a random port on localhost.
+    pub fn new_for_test() -> Self {
```
Comment (nagisa) [inline chain/network/src/tcp.rs:163]: Add `#[cfg(test)]`.

File: chain/network/src/tcp.rs [hunk 2]
```diff
@@ -118,15 +118,74 @@ impl Stream {
...
+impl std::ops::Deref for ListenerAddr {
+    type Target = std::net::SocketAddr;
+    fn deref(&self) -> &Self::Target {
+        &self.addr
+    }
+}
+
+impl ListenerAddr {
+    pub fn new(addr: std::net::SocketAddr) -> Self {
+        assert!(addr.port()!=0, "using an arbitrary port for the tcp::Listener is allowed only in tests and only via new_for_test() method");
```
Comment (nagisa) [inline chain/network/src/tcp.rs:158]: Did `rustfmt` leave it like this?

```suggestion
        assert!(
            addr.port() != 0, 
            "using an arbitrary port for the tcp::Listener is allowed only \
             in tests and only via new_for_test() method"
         );
```

File: chain/network/src/tcp.rs [hunk 3]
```diff
@@ -118,15 +118,74 @@ impl Stream {
...
+        Self { addr, guard: Arc::new(None) }
     }
 
+    /// TEST-ONLY: constructs a ListenerAddr owning a random port on localhost.
+    pub fn new_for_test() -> Self {
+        let guard = tokio::net::TcpSocket::new_v4().unwrap();
+        guard.set_reuseaddr(true).unwrap();
+        guard.set_reuseport(true).unwrap();
+        guard.bind("127.0.0.1:0".parse().unwrap()).unwrap();
```
Comment (nagisa) [inline chain/network/src/tcp.rs:167]: Shouldn’t this prefer `new_v6` and `("::1", 0)` or something? IIRC v6 sockets also bind on v4 and work transparently.

File: chain/network/src/tcp.rs [hunk 4]
```diff
@@ -118,15 +118,74 @@ impl Stream {
     }
 }
 
-pub(crate) struct Listener(tokio::net::TcpListener);
+/// In production code ListenerAddr is isomorphic to std::net::SocketAddr.
+/// In tests it additionally allows to "reserve" a random unused TCP port:
```
Comment (nagisa) [inline chain/network/src/tcp.rs:122]:
```suggestion
/// In production code ListenerAddr is isomorphic to std::net::SocketAddr.
///
/// In tests it additionally allows to "reserve" a random unused TCP port:
```

File: chain/network/src/tcp.rs [hunk 5]
```diff
@@ -118,15 +118,74 @@ impl Stream {
...
+}
 
-impl Listener {
-    // TODO(gprusak): this shouldn't be async. It is only
-    // because TcpListener accepts anything that asynchronously resolves to SocketAddr.
-    pub async fn bind(addr: std::net::SocketAddr) -> std::io::Result<Self> {
-        Ok(Self(tokio::net::TcpListener::bind(addr).await?))
+impl fmt::Debug for ListenerAddr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.addr.fmt(f)
```
Comment (nagisa) [inline chain/network/src/tcp.rs:145]: This discards information about the `ListenerAddr` wrapper – I think it is worthwhile to keep it, given the non-trivial additional behaviour added on top by this structure.

Something like this maybe:

```suggestion
        f.debug_struct("ListenerAddr")
            .field("addr", &self.addr)
            .field("has_guard", &self.guard.is_some())
            .finish()
```

or something?

File: chain/network/src/peer_manager/tests/routing.rs [hunk 6]
```diff
@@ -28,9 +28,19 @@ use std::collections::HashSet;
 use std::net::Ipv4Addr;
 use std::sync::Arc;
 
+fn abort_on_panic() {
```
Comment (nagisa) [inline chain/network/src/peer_manager/tests/routing.rs:31]: Hm, I think this would be better off as a separate PR. There some issues with this hook, which I don't believe need to block the port binding changes in this PR.

File: chain/network/src/peer_manager/testonly.rs [hunk 7]
```diff
@@ -136,7 +136,7 @@ impl ActorHandler {
     pub fn peer_info(&self) -> PeerInfo {
         PeerInfo {
             id: PeerId::new(self.cfg.node_key.public_key()),
-            addr: self.cfg.node_addr.clone(),
+            addr: self.cfg.node_addr.as_ref().map(|a| **a),
```
Comment (nagisa) [inline chain/network/src/peer_manager/testonly.rs:139]: I’m seeing a fair number of these around the code – would it make sense to add a convenience getter to the config?

**Review by nagisa — APPROVED** [review]
In principle the approach makes a lot of sense to me. I left some style nits around the PR, but overall I think this is a really welcome change we should land.

File: chain/network/src/tcp.rs (hunk 1 above)
Comment (pompon0) [reply chain/network/src/tcp.rs:163]: We have tests outside of the crate which use this.

File: chain/network/src/tcp.rs (hunk 2 above)
Comment (pompon0) [reply chain/network/src/tcp.rs:158]: cargo fmt did

File: chain/network/src/peer_manager/testonly.rs (hunk 7 above)
Comment (pompon0) [reply chain/network/src/peer_manager/testonly.rs:139]: I like it explicit, at least as of today.

File: chain/network/src/peer_manager/tests/routing.rs (hunk 6 above)
Comment (pompon0) [reply chain/network/src/peer_manager/tests/routing.rs:31]: moved to https://github.com/near/nearcore/pull/8439

File: chain/network/src/tcp.rs (hunk 3 above)
Comment (pompon0) [reply chain/network/src/tcp.rs:175]: migrated our codebase to use ::1 for localhost

File: chain/network/src/peer_manager/tests/routing.rs [hunk 8]
```diff
@@ -756,7 +760,7 @@ async fn blacklist_all() {
     tracing::info!(target:"test", "start two nodes with 0 blacklisting everything");
     let mut cfgs = make_configs(&chain, rng, 2, 2, true);
     cfgs[0].peer_store.blacklist =
-        [blacklist::Entry::from_ip(Ipv4Addr::LOCALHOST.into())].into_iter().collect();
+        [blacklist::Entry::from_ip(Ipv6Addr::LOCALHOST.into())].into_iter().collect();
```
Comment (nagisa) [inline chain/network/src/peer_manager/tests/routing.rs:763]: Wouldn’t this have worked either way, given that blacklist converts IPs to IPv6 anyway?

File: chain/network/src/tcp.rs (hunk 1 above)
Comment (nagisa) [reply chain/network/src/tcp.rs:171]: Hm, in that case it'd be worth considering adding a crate feature and using [`required-features` manifest field](https://doc.rust-lang.org/cargo/reference/features.html?highlight=required-features#required-features) to ensure tests have that feature enabled.

It would require `--feature=that-new-feature` to get those tests running though, which is not a great experience. Especially because AFAIU `--all-features` is just infeasible in nearcore (not great!) Probably fine to leave as-is for the time being then, I guess.

File: chain/network/src/test_utils.rs [hunk 9]
```diff
@@ -25,22 +25,22 @@ use tracing::debug;
 static OPENED_PORTS: Lazy<Mutex<HashSet<u16>>> = Lazy::new(|| Mutex::new(HashSet::new()));
 
 /// Returns available port.
-pub fn open_port() -> u16 {
+pub fn open_port() -> std::net::SocketAddr {
```
Comment (nagisa) [inline chain/network/src/test_utils.rs:28]: This still suffers from TOCTOU. Is there any reason this can’t return the newly added `ListenerAddr`?

File: chain/network/src/tcp.rs (hunk 1 above)
Comment (pompon0) [reply chain/network/src/tcp.rs:171]: ultimately I want this to be internal to near-network crate, but that will require migrating the remaining integration tests.

File: chain/network/src/test_utils.rs (hunk 9 above)
Comment (pompon0) [reply chain/network/src/test_utils.rs:28]: also we have tests which use the json-representation of the config, in which case the ListenerAddr representation of the address will be dropped. However I guess I could replace ListenerAddr with a global set (just like open_ports does) of guards for tcp ports allocated during the test. WDYT?

File: chain/network/src/peer_manager/tests/routing.rs (hunk 8 above)
Comment (nagisa) [reply chain/network/src/peer_manager/tests/routing.rs:763]: Hm, that’s a little scary, but you know this code better so I’ll leave the call here up to you.

File: chain/network/src/test_utils.rs (hunk 9 above)
Comment (nagisa) [reply chain/network/src/test_utils.rs:28]: I think its better to land as is right now and look into addressing it as a follow-up if you think it is a good use of our current resources.

**Comment by nagisa** [comment]
The sanity tests are still failing with what seems to be a relevant error:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ParseAccountError { kind: InvalidChar, char: Some((9, ':')) }', chain/network/src/network_protocol/peer.rs:86:53
```

**Review by nagisa — APPROVED** [review]
This makes a fair bit of sense to me, but I must admit that I did only a very light sweep over the test changes.

One concern  that comes to mind is the fact that this is at odds with `cargo test`. While we did migrate to `cargo nextest` in *CI*, I don't think developers are aware of this as much and may continue running `cargo test` locally – I know that at least I do.

Right now we are nowhere near hitting this problem, but if we do end up with enough tests to reserve 65k-or-so-ports, then the system would would basically just run out of available SockAddrs because there are no processes being terminated to clean up the reservations.

But that seems like something for the very-far-in-the-future-us to think about. Maybe by that time `cargo nextest` is used universally for other reasons.
